File: kirun-js/__tests__/engine/function/system/date/GetMinutesTest.ts
import { FunctionExecutionParameters, KIRunFunctionRepository, KIRunSchemaRepository, Namespaces } from "../../../../../src";
import { AbstractDateFunction } from "../../../../../src/engine/function/system/date/AbstractDateFunction";
import { DateFunctionRepository } from "../../../../../src/engine/function/system/date/DateFunctionRepository";


const dfr : DateFunctionRepository = new DateFunctionRepository();

const fep : FunctionExecutionParameters = new FunctionExecutionParameters(
    new KIRunFunctionRepository(),
    new KIRunSchemaRepository());

test('check for invalid dates', async () => {

    const getMinutesFunction = await dfr.find(Namespaces.DATE, 'GetMinutes');

    if (!getMinutesFunction) {
        throw new Error("Function not found");
    }

    fep.setArguments(new Map([[AbstractDateFunction.PARAMETER_DATE_NAME, '2029-02-00T06:04:18.073Z' ]]));

    await expect( () => getMinutesFunction.execute(fep)).rejects.toThrow();
    
});


test('check for invalid dates', async () => {

    const getMinutesFunction = await dfr.find(Namespaces.DATE, 'GetMinutes');

    if (!getMinutesFunction) {
        throw new Error("Function not found");
    }

    fep.setArguments(new Map([[AbstractDateFunction.PARAMETER_DATE_NAME, 'water' ]]));

    await expect( () => getMinutesFunction.execute(fep)).rejects.toThrow();
    
});

test('fetching for valid date' , async () => {

    const getMinutesFunction = await dfr.find(Namespaces.DATE, 'GetMinutes');

    if (!getMinutesFunction) {
        throw new Error("Function not found");
    }

    fep.setArguments(new Map([[AbstractDateFunction.PARAMETER_DATE_NAME, '2024-09-01T23:52:34.633-05:30' ]]));


    expect( (await getMinutesFunction.execute(fep)).allResults()[0].getResult().get(AbstractDateFunction.EVENT_RESULT_NAME)).toBe(22);

})

test('fetching for valid date' , async () => {

    const getMinutesFunction = await dfr.find(Namespaces.DATE, 'GetMinutes');

    if (!getMinutesFunction) {
        throw new Error("Function not found");
    }

    fep.setArguments(new Map([[AbstractDateFunction.PARAMETER_DATE_NAME, '2019-11-13T00:52:34.633Z' ]]));


    expect( (await getMinutesFunction.execute(fep)).allResults()[0].getResult().get(AbstractDateFunction.EVENT_RESULT_NAME)).toBe(52);

})


test('fetching for valid date' , async () => {

    const getMinutesFunction = await dfr.find(Namespaces.DATE, 'GetMinutes');

    if (!getMinutesFunction) {
        throw new Error("Function not found");
    }

    fep.setArguments(new Map([[AbstractDateFunction.PARAMETER_DATE_NAME, '2023-12-31T07:59:17.000-12:00' ]]));


    expect( (await getMinutesFunction.execute(fep)).allResults()[0].getResult().get(AbstractDateFunction.EVENT_RESULT_NAME)).toBe(59);

})

File: kirun-java/src/test/java/com/fincity/nocode/kirun/engine/function/system/date/GetMinutesTest.java
package com.fincity.nocode.kirun.engine.function.system.date;

import java.util.Map;

import org.junit.jupiter.api.Test;

import com.fincity.nocode.kirun.engine.namespaces.Namespaces;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveFunctionRepository;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveSchemaRepository;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.google.gson.JsonPrimitive;

import reactor.test.StepVerifier;

class GetMinutesTest {

    DateFunctionRepository dfr = new DateFunctionRepository();

    ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
            new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository());

    @Test
    void minutesSuccessTest1() {

        fep.setArguments(Map.of("isodate", new JsonPrimitive("2023-09-07T17:35:17.000Z")));

        StepVerifier.create(dfr.find(Namespaces.DATE, "GetMinutes")
                .flatMap(e -> e.execute(fep)))
                .expectNextMatches(
                        res -> res.next().getResult().get("minutes").getAsInt() == 5)
                .verifyComplete();

        fep.setArguments(Map.of("isodate", new JsonPrimitive("2023-10-19T06:44:11.615Z")));

        StepVerifier.create(dfr.find(Namespaces.DATE, "GetMinutes")
                .flatMap(e -> e.execute(fep)))
                .expectNextMatches(
                        res -> res.next().getResult().get("minutes").getAsInt() == 14)
                .verifyComplete();

        fep.setArguments(Map.of("isodate", new JsonPrimitive("1970-01-20T15:58:57.561-12:31")));

        StepVerifier.create(dfr.find(Namespaces.DATE, "GetMinutes")
                .flatMap(e -> e.execute(fep)))
                .expectNextMatches(
                        res -> res.next().getResult().get("minutes").getAsInt() == 59)
                .verifyComplete();

        fep.setArguments(Map.of("isodate", new JsonPrimitive("2507-08-08T11:41:50.000+09:00")));

        StepVerifier.create(dfr.find(Namespaces.DATE, "GetMinutes")
                .flatMap(e -> e.execute(fep)))
                .expectNextMatches(
                        res -> res.next().getResult().get("minutes").getAsInt() == 11)
                .verifyComplete();

        fep.setArguments(Map.of("isodate", new JsonPrimitive("1970-01-20T15:13:51.001+12:01")));

        StepVerifier.create(dfr.find(Namespaces.DATE, "GetMinutes")
                .flatMap(e -> e.execute(fep)))
                .expectNextMatches(
                        res -> res.next().getResult().get("minutes").getAsInt() == 42)
                .verifyComplete();

        fep.setArguments(Map.of("isodate", new JsonPrimitive("1970-01-20T15:13:51.200-12:01")));

        StepVerifier.create(dfr.find(Namespaces.DATE, "GetMinutes")
                .flatMap(e -> e.execute(fep)))
                .expectNextMatches(
                        res -> res.next().getResult().get("minutes").getAsInt() == 44)
                .verifyComplete();

    }

    @Test
    void minutesFailTest1() {

        fep.setArguments(Map.of("isodate", new JsonPrimitive("2023-09-7T07:35:17.000Z")));

        StepVerifier.create(dfr.find(Namespaces.DATE, "GetMinutes")
                .flatMap(e -> e.execute(fep)))
                .expectErrorMessage("Please provide the valid iso date.")
                .verify();

        fep.setArguments(Map.of("isodate", new JsonPrimitive(false)));

        StepVerifier.create(dfr.find(Namespaces.DATE, "GetMinutes")
                .flatMap(e -> e.execute(fep)))
                .expectError()
                .verify();

        fep.setArguments(Map.of("isodate", new JsonPrimitive("2023-10-19T23:84:11.615Z")));

        StepVerifier.create(dfr.find(Namespaces.DATE, "GetMinutes")
                .flatMap(e -> e.execute(fep)))
                .expectErrorMessage("Please provide the valid iso date.")
                .verify();
    }

}
